spring3.0.2以上版本+nacos +druid，配置nacos动态刷新后刷新任意一个无关的key都会导致druid连接断开并无法重连。

配置nacos动态刷新后，更新nacos上任意key值，会导致使用druid的服务连接不上数据库报错，大致信息如下：

`org.springframework.boot.context.properties.ConfigurationPropertiesBindException: Error creating bean with name 'dataSource': Could not bind properties to 'DataSource' : prefix=spring.datasource.druid, ignoreInvalidFields=false, ignoreUnknownFields=true
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.bind(ConfigurationPropertiesBindingPostProcessor.java:99)
      .
      .
      .
Caused by: com.alibaba.druid.pool.DataSourceClosedException: dataSource already closed at Sat Jan 13 16:55:12 CST 2024
	at com.alibaba.druid.pool.DruidDataSource.getConnectionInternal(DruidDataSource.java:1676)
	at com.alibaba.druid.pool.DruidDataSource.getConnectionDirect(DruidDataSource.java:1504)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1484)
	at com.alibaba.druid.pool.DruidDataSource.getConnection(DruidDataSource.java:1469)
	... 61 common frames omitted`

初步排查是因为springboot重新绑定是调用调用bean的destroy方法之后重新初始化 bean，

> appContext.getAutowireCapableBeanFactory().destroyBean(bean);
				appContext.getAutowireCapableBeanFactory().initializeBean(bean, name);

对于druid，destroy后，druid连接池关闭（closed= true）然后进行初始化，包括bean的属性bind；
对于复杂属性，[spirngboot3.0.2版本修改了构造器绑定的逻辑](https://github.com/spring-projects/spring-boot/commit/84b13f07488ed220ce1ab8c5148306f72dde727e)，新逻辑会先调用bean的get方法判断属性是否已经存在来确定是否使用构造器绑定来初始化属性，在判断DruidDataSourceWrapper的connection属性时，由于closed = true ，getConnection()方法抛出异常，导致问题出现。

--- a/druid-spring-boot-3-starter/src/main/java/com/alibaba/druid/spring/boot3/autoconfigure/DruidDataSourceWrapper.java
+++ b/druid-spring-boot-3-starter/src/main/java/com/alibaba/druid/spring/boot3/autoconfigure/DruidDataSourceWrapper.java
@@ -17,7 +17,6 @@ package com.alibaba.druid.spring.boot3.autoconfigure;
 
 import com.alibaba.druid.filter.Filter;
 import com.alibaba.druid.pool.DruidDataSource;
-import org.springframework.beans.factory.DisposableBean;
 import org.springframework.beans.factory.InitializingBean;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
@@ -29,7 +28,7 @@ import java.util.List;
  * @author lihengming [89921218@qq.com]
  */
 @ConfigurationProperties("spring.datasource.druid")
-public class DruidDataSourceWrapper extends DruidDataSource implements InitializingBean, DisposableBean {
+public class DruidDataSourceWrapper extends DruidDataSource implements InitializingBean {
     @Autowired
     private DataSourceProperties basicProperties;
 
@@ -72,10 +71,5 @@ public class DruidDataSourceWrapper extends DruidDataSource implements Initializ
         } catch (IllegalArgumentException ignore) {
             super.maxEvictableIdleTimeMillis = maxEvictableIdleTimeMillis;
         }
-    }
-
-    @Override
-    public void destroy() throws Exception {
-        close();
     }
 }

--- a/druid-spring-boot-starter/src/main/java/com/alibaba/druid/spring/boot/autoconfigure/DruidDataSourceWrapper.java
+++ b/druid-spring-boot-starter/src/main/java/com/alibaba/druid/spring/boot/autoconfigure/DruidDataSourceWrapper.java
@@ -17,7 +17,6 @@ package com.alibaba.druid.spring.boot.autoconfigure;
 
 import com.alibaba.druid.filter.Filter;
 import com.alibaba.druid.pool.DruidDataSource;
-import org.springframework.beans.factory.DisposableBean;
 import org.springframework.beans.factory.InitializingBean;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
@@ -29,7 +28,7 @@ import java.util.List;
  * @author lihengming [89921218@qq.com]
  */
 @ConfigurationProperties("spring.datasource.druid")
-public class DruidDataSourceWrapper extends DruidDataSource implements InitializingBean, DisposableBean {
+public class DruidDataSourceWrapper extends DruidDataSource implements InitializingBean {
     @Autowired
     private DataSourceProperties basicProperties;
 
@@ -72,10 +71,5 @@ public class DruidDataSourceWrapper extends DruidDataSource implements Initializ
         } catch (IllegalArgumentException ignore) {
             super.maxEvictableIdleTimeMillis = maxEvictableIdleTimeMillis;
         }
-    }
-
-    @Override
-    public void destroy() throws Exception {
-        close();
     }
 }

